python/doc/examples/numerical_methods/general_methods/plot_estimate_integral_iterated_quadrature.py

```diff
@@ -1,48 +1,272 @@
-#    \, f(x_0, \dots, x_{n-1})\mathrm{d}{x_{n-1}}\dots\mathrm{d}{x_0}
+#    I_f = \int_{a}^{b}\, \int_{l_1(x_0)}^{u_1(x_0)}\, \int_{l_2(x_0, x_1)}^{u_2(x_0,x_1)}\, \int_{l_{d-1}(x_0, \dots, x_{d-2})}^{u_{d-1}(x_0, \dots, x_{d-2})}
+#    \, f(x_0, \dots, x_{d-1})\mathrm{d}{x_{d-1}}\dots\mathrm{d}{x_0}
 #
-# with the iterated quadrature algorithm.
+# using an iterated quadrature algorithm :class:`~openturns.IteratedQuadrature`, based on the Gauss-Kronrod
+# :class:`~openturns.GaussKronrod` 1d quadrature algorithm.
 
-# %%
 import openturns as ot
-import openturns.viewer as viewer
-from matplotlib import pylab as plt
+import openturns.viewer as otv
 import math as m
 
-ot.Log.Show(ot.Log.NONE)
+# %%
+# To get better colours for the function iso-lines.
+ot.ResourceMap.SetAsString("Contour-DefaultColorMapNorm", "rank")
+
+# %%
+# Case 1: The integrand function presents a lot of discontinuities.
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# We consider the function:
+#
+# .. math::
+#    f : \begin{array}{lcl}
+#          \Rset^2 & \rightarrow & \Rset \\
+#          (x,y) & \rightarrow & 1_{{x^2+y^2 \leq 4}}(x,y)
+#          \end{array}
+#
+# the domain :math:`\set{D} =[-2,2]\times [-2,2]` and the integral:
+#
+# .. math::
+#    I = \int_{\set{D}} f(x,y)\,dxdy.
+#
+
+# %%
+# We first define the integrand :math:`f` and the domain :math:`\set{D}`.
+
+
+def kernel1(x):
+    if x[0]**2 + x[1]**2 <= 4:
+        return [1.0]
+    else:
+        return [0.0]
+
+
+integrand = ot.PythonFunction(2, 1, kernel1)
+domain = ot.Interval([-2.0, -2.0], [2.0, 2.0])
+
+# %%
+# We draw the iso-lines of the integrand function which is
+# constant equal to 1 inside the cercle and equal to 0.0 outside.
+g = integrand.draw([-3.0, -3.0], [3.0, 3.0])
+g.setTitle(r'Function $f(x,y) = 1_{\{x+2+y^2 \leq 4\}}(x,y)$')
+g.setXTitle('x')
+g.setYTitle('y')
+
+# %%
+# We compute the integral using an  :class:`~openturns.IteratedQuadrature`
+# algorithm with
+# the :class:`~openturns.GaussKronrodRule`  *G11K23*, which means that we use
+# on each subinterval:
+#
+# - 11 points for the Gauss approximations,
+# - 23 points for the Kronrod approximation including the 11 previous ones.
+#
+# We limit the number of subintervals and we  define a maximum error between both approximations:
+# when the difference between both approximations is lower, a new subdivision is performed.
+maximumSubIntervals = 32
+maximumError = 1e-4
+GKRule = ot.GaussKronrodRule(ot.GaussKronrodRule.G1K3)
+integration_algo = ot.IteratedQuadrature(ot.GaussKronrod(maximumSubIntervals, maximumError, GKRule))
+
+# %%
+# To get the nodes used to approximate the integral, we need to use a :class:`~openturns.MemoizeFunction`
+# that stores all the calls to the function.
+integrand_memoized = ot.MemoizeFunction(integrand)
+I_value = integration_algo.integrate(integrand_memoized, domain)
+print('I = ', I_value[0])
```

Comment:
You could also give the exact value ($4\pi$ here)

Reply:
yes